br/pkg/storage/gcs.go

```diff
@@ -465,3 +468,83 @@ func (r *gcsObjectReader) Seek(offset int64, whence int) (int64, error) {
 func (r *gcsObjectReader) GetFileSize() (int64, error) {
 	return r.totalSize, nil
 }
+
+type gcsMultiWriter struct {
+	eg     errgroup.Group
+	prefix string
+	seq    int
+	s      *GCSStorage
+}
+
+// Write writes data to the GCS object.
+func (w *gcsMultiWriter) Write(ctx context.Context, data []byte) (int, error) {
+	object := w.s.objectName(w.prefix + strconv.Itoa(w.seq))
+	w.seq++
+	w.eg.Go(
+		func() error {
+			wc := w.s.bucket.Object(object).NewWriter(ctx)
+			wc.StorageClass = w.s.gcs.StorageClass
+			wc.PredefinedACL = w.s.gcs.PredefinedAcl
+			_, _ = wc.Write(data)
+			return wc.Close()
+		})
+	return 0, nil
+}
+
+// Close closes the writer and returns the error if exists.
+func (w *gcsMultiWriter) Close(ctx context.Context) error {
+	err := w.eg.Wait()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	ts := time.Now()
+	batch := 30
+	headObject := w.s.bucket.Object(w.s.objectName(w.prefix))
+	var tempObject *storage.ObjectHandle
+	objectsToConcat := make([]*storage.ObjectHandle, 0, batch)
+	allTempObjects := make([]*storage.ObjectHandle, 0, batch)
+	for i := 0; i < w.seq; i += batch {
+		if i+batch < w.seq {
+			tempObject = w.s.bucket.Object(w.s.objectName(w.prefix + "temp" + "_" + strconv.Itoa(i)))
+		} else {
+			tempObject = headObject
+		}
+		for j := i; j < i+batch && j < w.seq; j++ {
+			objectsToConcat = append(objectsToConcat, w.s.bucket.Object(w.s.objectName(w.prefix+strconv.Itoa(j))))
+			allTempObjects = append(allTempObjects, w.s.bucket.Object(w.s.objectName(w.prefix+strconv.Itoa(j))))
+		}
+		_, err := tempObject.ComposerFrom(objectsToConcat...).Run(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if i+batch < w.seq {
+			objectsToConcat = objectsToConcat[:0]
+			objectsToConcat = append(objectsToConcat, tempObject)
+			allTempObjects = append(allTempObjects, tempObject)
+		}
+	}
+
+	log.FromContext(ctx).Info("concat files", zap.Int("count", w.seq), zap.Duration("cost", time.Since(ts)))
+	ts = time.Now()
+
+	for _, oj := range allTempObjects {
+		err := oj.Delete(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	log.FromContext(ctx).Info("delete temporary files", zap.Int("count", w.seq), zap.Duration("cost", time.Since(ts)))
```

Comment:
what's the time cost of it? Do we need to run deletes asynchronized after `Close` returns?